autogpt_platform/backend/backend/server/rest_api.py

```diff
@@ -481,40 +492,79 @@ async def update_graph(
         latest_version_graph = next(
             v for v in existing_versions if v.version == latest_version_number
         )
+        current_active_version = next(
+            (v for v in existing_versions if v.is_active), None
+        )
         if latest_version_graph.is_template != graph.is_template:
             raise HTTPException(
                 400, detail="Changing is_template on an existing graph is forbidden"
             )
         graph.is_active = not graph.is_template
+        graph = graph_db.graph_from_creatable(graph, user_id)
         graph.reassign_ids()
 
         new_graph_version = await graph_db.create_graph(graph, user_id=user_id)
 
         if new_graph_version.is_active:
+
+            def get_credentials(credentials_id: str) -> "Credentials | None":
+                return self.integration_creds_manager.get(user_id, credentials_id)
+
+            # Handle activation of the new graph first to ensure continuity
+            new_graph_version = await on_graph_activate(
+                new_graph_version,
+                get_credentials=get_credentials,
+            )
             # Ensure new version is the only active version
             await graph_db.set_graph_active_version(
                 graph_id=graph_id, version=new_graph_version.version, user_id=user_id
             )
+            if current_active_version:
+                # Handle deactivation of the previously active version
+                await on_graph_deactivate(
+                    current_active_version,
+                    get_credentials=get_credentials,
+                )
 
         return new_graph_version
 
-    @classmethod
     async def set_graph_active_version(
-        cls,
+        self,
         graph_id: str,
         request_body: SetGraphActiveVersion,
         user_id: Annotated[str, Depends(get_user_id)],
     ):
         new_active_version = request_body.active_graph_version
-        if not await graph_db.get_graph(graph_id, new_active_version, user_id=user_id):
+        new_active_graph = await graph_db.get_graph(
+            graph_id, new_active_version, user_id=user_id
+        )
+        if not new_active_graph:
             raise HTTPException(
                 404, f"Graph #{graph_id} v{new_active_version} not found"
             )
+
+        current_active_graph = await graph_db.get_graph(graph_id, user_id=user_id)
+
+        def get_credentials(credentials_id: str) -> "Credentials | None":
+            return self.integration_creds_manager.get(user_id, credentials_id)
+
+        # Handle activation of the new graph first to ensure continuity
+        await on_graph_activate(
```

Comment:
Why do we need to do this if the graph is already active ? Why are we only skipping on_graph_deactivate in such case.